docs: Tutorial for ODSC presentation (#703)
awslabs/syne-tune: docs/source/tutorials/odsc_tutorial/getting_started.rst

Getting Started with Hyperparameter Tuning
==========================================

In this section, you will learn what is needed to get hyperparameter tuning up

> and running. We will look at an example, where a deep learning language model

Reviewer: ```suggestion
and running. We will look at an example where a deep learning language model
```

What is Hyperparameter Tuning?
------------------------------

When solving a business problem with machine learning, there are parts which can
be automated by spending compute resources, and other parts require human
expert attention and choices to be made. By automating some of the more tedious
parts of the latter, hyperparameter tuning shifts the needle between these cost
factors. Like any other smart tool, it saves you time to concentrate on where your
strengths really lie, and where you can create the most value.

At a high level, hyperparameter tuning finds *configurations* of a system which
optimize a *target metric* (or several ones, as we will see later). We can try
any configuration from a *configuration space*, but each evaluation of the system
has a *cost* and takes time. The main challenge of hyperparameter tuning is to
run as few trials as possible, so that total costs are minimal. Also, if
possible, trials should be run in parallel, so that the total experiment time
is minimal.

In this tutorial, we will mostly
be focussed on making decisions and tuning free parameters in the context of
*training machine learning models on data*, so their predictions can be used as

> part of a solution to a business problem. There are many other steps from
> initial need to deployed solution, such as understanding business requirements,

Reviewer: ```suggestion
part of a solution to a business problem. There are many other steps between the
initial need and the deployed solution, such as understanding business requirements,
```

  gradient descent on weight and biases of a neural network model), *given*
  a configuration (e.g., what learning rate is used, what batch size, etc.).
  Then, the trained model is evaluated on another part of the data (validation
  data, disjoint from training data), giving rise to a quality metric (e.g.,
  validation error, AUC, F1), or even several ones. For small datasets, we can
  also use cross-validation, by repeating training and evaluation on a
  number of different splits, reporting the average of validation metrics.
* This metric value (or values) is the response of the system to a
  configuration. Note that the response is stochastic: if we run again with
  the same configuration, we may get a different value. This is because training
  has random elements (e.g., initial weights are sampled, ordering of training
  data).

Enough high level and definitions, let us dive into an example.

Annotating a Training Script
----------------------------

First, we need a script to execute a trial, by training a model and evaluating it.
Since training models is bread and butter to machine learners, you will have no

> problem to come up with one. We start with an example

Reviewer: ```suggestion
problem to come up with one. We start with an example:
```
